### PR TITLE
Init

### PR DESCRIPTION
**Pull request by ajlee21**
This PR initializes my cycleGAN repo

There are two main notebooks:

1.  0_process_data.ipynb is exploring the raw data by plotting the raw gene expression data on UMAP dimensions and coloring my metadata

2.  1_train_cycleGAN.ipynb is training cycleGAN on matched input and target domains using the entire Pa compendium as the two domains.

Next PR will verify the loss function and why there is a spike at the last epoch.  I will also update the input and target domains and look to see what the generator is learning.

### CODE REVIEW

File: env/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+torch>=0.4.0
```
Comment (danich1) [inline env/requirements.txt]: Is this file needed? You have a yml file that contains all the packages to install. Shouldn't that suffice?

File: scripts/functions/cyclegan_transcript.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+if cuda:
+    G_AB = G_AB.cuda()
+    G_BA = G_BA.cuda()
+    D_A = D_A.cuda()
+    D_B = D_B.cuda()
+    criterion_GAN.cuda()
+    criterion_cycle.cuda()
+    criterion_identity.cuda()
+
+if opt.epoch != 0:
```
Comment (danich1) [inline scripts/functions/cyclegan_transcript.py]: I know you are using pre-written code, but to make life easier for your future self I would change opt.epoch -> opt.epoch_resume or opt.preload_epoch.

File: scripts/functions/cyclegan_transcript.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        # -----------------------
+        #  Train Discriminator A
+        # -----------------------
+
+        optimizer_D_A.zero_grad()
+
+        # Real loss
+        loss_real = criterion_GAN(D_A(real_A), valid)
+        # Fake loss (on batch of previously generated samples)
+        fake_A_ = fake_A_buffer.push_and_pop(fake_A)
```
Comment (danich1) [inline scripts/functions/cyclegan_transcript.py:256]: Do you save any of the generated output from the generator? If no I do recommend you save it next time you look into this.

File: scripts/functions/datasets_transcript.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+from torch.utils.data import Dataset
+from PIL import Image
+import torchvision.transforms as transforms
+
+
+class GeneExpressionDataset(Dataset):
+    def __init__(self, root, mode="train"):
+        # Training data loader
+        data_file = os.path.join("../../data/%s" % opt.dataset_name +
```
Comment (danich1) [inline scripts/functions/datasets_transcript.py]: Are you using python 3.6?  if so I recommend using the new string formatting:
```python
data_file = f"../../data/{opt.dataset_name}/train/A/all-pseudomonas-gene-normalized.zip"
```

File: scripts/nbconverted/0_process_data.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+# 
+# from [Exploration, normalization, and summaries of high density oligonucleotide array probe level data](https://academic.oup.com/biostatistics/article/4/2/249/245074)
+# 
+# **Alternative normalization methods**
+
+# In[2]:
+
+
+# Load arguments
+data_file = os.path.join(os.path.dirname(os.getcwd()), "data", "pseudomonas", "Pa_compendium_02.22.2014.pcl")
```
Comment (danich1) [inline scripts/nbconverted/0_process_data.py]: Might want to consider 3.6 new string formatting:
```python
data_file = f"{os.path.dir(os.getcwd())}/data/pseudomonas/Pa_compendium_02.22.2014.pcl"
```

File: scripts/nbconverted/0_process_data.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+metadata_selected.head(10)
+
+
+# ## Plot RMA normalized data
+
+# In[11]:
+
+
+data_labeled = data.merge(metadata_selected_labeled, left_index=True, right_index=True, how='inner')
```
Comment (danich1) [inline scripts/nbconverted/0_process_data.py]: Change to:
```python
data_labeled = data.merge(
   metadata_selected_labeled,
   left_index=True,
   right_index=True,
   how='inner'
 )
```

File: scripts/nbconverted/1_train_cycleGAN.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+from numpy.random import seed
+randomState = 123
+seed(randomState)
+
+
+# In[2]:
+
+
+# Load arguments
+data_file = os.path.join(
```
Comment (danich1) [inline scripts/nbconverted/1_train_cycleGAN.py]: Look at string format comment above.

File: scripts/functions/models_transcript.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+
+class GeneratorResNet(nn.Module):
+    def __init__(self, input_shape, hidden_dim, output_dim, num_residual_blocks):
+        super(GeneratorResNet, self).__init__()
+
+        input_dim = input_shape[1]
+
+        # Encoder (Downsampling)
+        model = [
```
Comment (danich1) [inline scripts/functions/models_transcript.py:46]: For your model do you need to do this down sampling/resnet block stuff? Could you get away with a simple fully connected network such as:
```
nn.Linear(input_dim, encoding_dim),
nn.ReLU(),
nn.Linear(encoding_dim, hidden_dim),
nn.ReLU(),
nn.Linear(hidden_dim, input_dim)
```
or something of that variety?

**Review by danich1 — APPROVED** [review]
Awesome PR. Looks like cycleGANs could product interesting results. 
1. Just made some minor code comments
2. If you aren't saving the Generator output I highly recommend you do.
3. Next time you run this code plot the generator loss and the discriminator loss on the same plot. Also for the two plots which generator and discriminator are you plotting (Gen A, Disc A/ Gen B, Disc B)?

Gonna pre-approve given you make those changes and answer a few questions.

File: env/requirements.txt (hunk 1 above)
Comment (ajlee21) [reply env/requirements.txt]: you're right, I'll remove this.  This is leftover from the original cycleGAN repo that was cloned i believe

File: scripts/functions/cyclegan_transcript.py (hunk 2 above)
Comment (ajlee21) [reply scripts/functions/cyclegan_transcript.py]: Thank you for the tip.  I definitely need to spend time to more carefully walk through the script and the adjustments I have made to make sure they are correct, just haven't had time yet.

File: scripts/functions/cyclegan_transcript.py (hunk 3 above)
Comment (ajlee21) [reply scripts/functions/cyclegan_transcript.py:256]: Very good point.  This will definitely be key in evaluating if this method is "working" the way we think it should.  Currently I am not.  This PR was just to get cycleGAN running on gene expression data.  Future PR will save the generator model and the transformed data as well.

File: scripts/functions/datasets_transcript.py (hunk 4 above)
Comment (ajlee21) [reply scripts/functions/datasets_transcript.py]: It is using py3.6  So is this new syntax for this version?  It does look cleaner, though I do think its easier to read if I can break up the lines which I don't think I can do here...

File: scripts/nbconverted/0_process_data.py (hunk 5 above)
Comment (ajlee21) [reply scripts/nbconverted/0_process_data.py]: done

File: scripts/nbconverted/0_process_data.py (hunk 6 above)
Comment (ajlee21) [reply scripts/nbconverted/0_process_data.py]: Done.  Definitely like this new format, its so much easier to read.  Unfortunately I'm still not consistently doing it sometimes.  Thanks for catching this

File: scripts/nbconverted/1_train_cycleGAN.py (hunk 7 above)
Comment (ajlee21) [reply scripts/nbconverted/1_train_cycleGAN.py]: Done

File: scripts/functions/models_transcript.py (hunk 8 above)
Comment (ajlee21) [reply scripts/functions/models_transcript.py:46]: So this is misleading.  I did change the script to use a fully connected block instead of this resnet, but I kept original function names and comments.  But there is an encoder --> fully connected layer-->decoder
